### PR TITLE
RavenDB-13293: Add option to pass node to RestoreBackupOperation

### PR DESCRIPTION
**Pull request by garayx**

### CODE REVIEW

File: test/SlowTests/Issues/RavenDB_13293.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                    });
+                }
+
+                foreach (var myBackup in myBackupsList)
+                {
+                    await store.Maintenance.SendAsync(new StartBackupOperation(true, myBackup.BackupTaskId));
+                    var backupOperation = new GetPeriodicBackupStatusOperation(myBackup.BackupTaskId);
+                    var getPeriodicBackupResult = store.Maintenance.Send(backupOperation);
+
+                    SpinWait.SpinUntil(() =>
```
Comment (grisha-kotler) [inline test/SlowTests/Issues/RavenDB_13293.cs]: WaitForValue

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

**Review by ayende — APPROVED** [review]

**Review by grisha-kotler — APPROVED** [review]

File: src/Raven.Client/Exceptions/RequestedNodeUnavailableException.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+
+namespace Raven.Client.Exceptions
+{
+    public class RequestedNodeUnavailableException : Exception
```
Comment (ppekrol) [inline src/Raven.Client/Exceptions/RequestedNodeUnavailableException.cs:5]: needs to inherit from RavenException?

File: src/Raven.Client/ServerWide/Operations/ServerWideOperation.cs [hunk 3]
```diff
@@ -6,15 +6,16 @@ namespace Raven.Client.ServerWide.Operations
 {
     public class ServerWideOperation : Operation
     {
-        public ServerWideOperation(RequestExecutor requestExecutor, DocumentConventions conventions, long id)
+        public ServerWideOperation(RequestExecutor requestExecutor, DocumentConventions conventions, long id, string commandSelectedNodeTagForRequest = null)
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/ServerWideOperation.cs]: rename to `nodeTag`

File: src/Raven.Client/Http/RavenCommand.cs [hunk 4]
```diff
@@ -44,6 +44,7 @@ public abstract class RavenCommand<TResult>
         public TimeSpan? Timeout { get; protected set; }
         public bool CanCache { get; protected set; }
         public bool CanCacheAggressively { get; protected set; }
+        public string SelectedNodeTagForRequest { get; protected set; }
```
Comment (ppekrol) [inline src/Raven.Client/Http/RavenCommand.cs]: rename to `SelectedNodeTag`

File: src/Raven.Client/Http/NodeSelector.cs [hunk 5]
```diff
@@ -60,6 +60,29 @@ public bool OnUpdateTopology(Topology topology, bool forceUpdate = false)
             return true;
         }
 
+        public (int Index, ServerNode Node) GetRequestedNode(string nodeTag)
```
Comment (ppekrol) [inline src/Raven.Client/Http/NodeSelector.cs]: internal

File: src/Raven.Client/Http/NodeSelector.cs [hunk 6]
```diff
@@ -60,6 +60,29 @@ public bool OnUpdateTopology(Topology topology, bool forceUpdate = false)
...
+        {
+            var state = _state;
```
Comment (ppekrol) [inline src/Raven.Client/Http/NodeSelector.cs:65]: This code should be moved to NodeSelector, I think?

**Review by ppekrol — CHANGES_REQUESTED** [review]
This only solved the issue for ServerWideOperations, but things like Patching are using ChangesAPI and we need to take care of this also.

File: src/Raven.Client/ServerWide/Operations/ServerWideOperation.cs (hunk 3 above)
Comment (garayx) [reply src/Raven.Client/ServerWide/Operations/ServerWideOperation.cs]: done

File: src/Raven.Client/Http/RavenCommand.cs (hunk 4 above)
Comment (garayx) [reply src/Raven.Client/Http/RavenCommand.cs]: done

File: src/Raven.Client/Http/NodeSelector.cs (hunk 5 above)
Comment (garayx) [reply src/Raven.Client/Http/NodeSelector.cs]: done

File: src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs [hunk 7]
```diff
@@ -1597,13 +1597,18 @@ public static void WriteDocumentPropertiesWithoutMetadata(this BlittableJsonText
...
+        public static void WriteOperationIdAndNodeTag(this BlittableJsonTextWriter writer, JsonOperationContext context, long operationId, string nodeTag)
         {
             writer.WriteStartObject();
 
             writer.WritePropertyName("OperationId");
             writer.WriteInteger(operationId);
 
+            writer.WriteComma();
+
+            writer.WritePropertyName("OperationNodeTag");
```
Comment (ppekrol) [inline src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs]: Let's use nameof

File: src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs [hunk 8]
```diff
@@ -1597,13 +1597,18 @@ public static void WriteDocumentPropertiesWithoutMetadata(this BlittableJsonText
             }
         }
 
-        public static void WriteOperationId(this BlittableJsonTextWriter writer, JsonOperationContext context, long operationId)
+        public static void WriteOperationIdAndNodeTag(this BlittableJsonTextWriter writer, JsonOperationContext context, long operationId, string nodeTag)
         {
             writer.WriteStartObject();
 
             writer.WritePropertyName("OperationId");
```
Comment (ppekrol) [inline src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs]: Let's use nameof

File: src/Raven.Client/Http/RavenCommand.cs [hunk 9]
```diff
@@ -44,7 +44,7 @@ public abstract class RavenCommand<TResult>
         public TimeSpan? Timeout { get; protected set; }
         public bool CanCache { get; protected set; }
         public bool CanCacheAggressively { get; protected set; }
-        public string SelectedNodeTagForRequest { get; protected set; }
+        public string SelectedNodeTag { get; protected set; }
```
Comment (ppekrol) [inline src/Raven.Client/Http/RavenCommand.cs:47]: Do we still need it?

File: src/Raven.Client/Documents/Operations/Operation.cs [hunk 10]
```diff
@@ -30,15 +30,17 @@ public class Operation : IObserver<OperationStatusChange>
         internal long Id => _id;
 
         public OperationStatusFetchMode StatusFetchMode { get; protected set; }
+        public string NodeTag { get; protected set; }
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Operations/Operation.cs]: `internal readonly string NodeTag`

File: src/Raven.Client/ServerWide/Operations/RestoreBackupOperation.cs [hunk 11]
```diff
@@ -13,10 +13,12 @@ namespace Raven.Client.ServerWide.Operations
     public class RestoreBackupOperation : IServerOperation<OperationIdResult>
     {
         private readonly RestoreBackupConfiguration _restoreConfiguration;
+        public string NodeTag;
 
         public RestoreBackupOperation(RestoreBackupConfiguration restoreConfiguration)
         {
             _restoreConfiguration = restoreConfiguration;
+            NodeTag = restoreConfiguration.NodeTag;
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/RestoreBackupOperation.cs]: I think that NodeTag should not be part of `RestoreBackupConfiguration`. It contains data that we are sending to the server and this is not related.

Comment (garayx) [reply src/Raven.Client/ServerWide/Operations/RestoreBackupOperation.cs]: done

File: src/Raven.Client/Documents/Operations/Operation.cs (hunk 10 above)
Comment (garayx) [reply src/Raven.Client/Documents/Operations/Operation.cs]: done

File: src/Raven.Client/Documents/Operations/GetOperationStateOperation.cs [hunk 12]
```diff
@@ -8,15 +8,22 @@ namespace Raven.Client.Documents.Operations
...
     {
         private readonly long _id;
+        private readonly string _nodeTag;
 
         public GetOperationStateOperation(long id)
         {
             _id = id;
         }
 
+        public GetOperationStateOperation(long id, string node)
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Operations/GetOperationStateOperation.cs]: nodeTag

File: src/Raven.Client/Documents/Operations/Operation.cs [hunk 13]
```diff
@@ -193,9 +195,9 @@ protected async Task FetchOperationStatus()
             });
         }
 
-        protected virtual RavenCommand<OperationState> GetOperationStateCommand(DocumentConventions conventions, long id)
+        protected virtual RavenCommand<OperationState> GetOperationStateCommand(DocumentConventions conventions, long id, string nodeTag = null)
         {
-            return new GetOperationStateOperation.GetOperationStateCommand(conventions, id);
+            return new GetOperationStateOperation.GetOperationStateCommand(conventions, id, nodeTag);
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Operations/Operation.cs:200]: we also need to handle ChangesAPI part

File: src/Raven.Client/Documents/Operations/GetOperationStateOperation.cs (hunk 12 above)
Comment (garayx) [reply src/Raven.Client/Documents/Operations/GetOperationStateOperation.cs]: done

File: src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs (hunk 7 above)
Comment (garayx) [reply src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs]: done

File: src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs (hunk 8 above)
Comment (garayx) [reply src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs]: done

File: src/Raven.Client/Http/RavenCommand.cs (hunk 9 above)
Comment (garayx) [reply src/Raven.Client/Http/RavenCommand.cs:47]: yes we use this to explicitly select node to execute the command

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 14]
```diff
@@ -1724,7 +1733,8 @@ public string HardResetToNewCluster(string nodeTag = "A")
                     },
                     new Dictionary<string, string>(),
                     new Dictionary<string, string>(),
-                    lastNode
+                    lastNode,
+                    -1
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/RachisConsensus.cs]: I think we do need to notify the client about cluster topology change. So we probably should set it to a higher value here.

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 15]
```diff
@@ -1753,7 +1763,8 @@ public void HardResetToPassive(string topologyId = null)
                     },
                     new Dictionary<string, string>(),
                     new Dictionary<string, string>(),
-                    string.Empty
+                    string.Empty,
+                    -1
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/RachisConsensus.cs]: same.

File: src/Raven.Server/Rachis/RachisConsensus.cs [hunk 16]
```diff
@@ -269,6 +269,7 @@ private set
         private readonly List<IDisposable> _disposables = new List<IDisposable>();
 
         public long CurrentTerm { get; private set; }
+        public long LastTopologyEtag { get; private set; }
```
Comment (karmeli87) [inline src/Raven.Server/Rachis/RachisConsensus.cs]: do we still need that?

File: src/Raven.Server/Rachis/RachisConsensus.cs (hunk 14 above)
Comment (garayx) [reply src/Raven.Server/Rachis/RachisConsensus.cs]: done

File: src/Raven.Server/Rachis/RachisConsensus.cs (hunk 15 above)
Comment (garayx) [reply src/Raven.Server/Rachis/RachisConsensus.cs]: done

File: src/Raven.Server/Rachis/RachisConsensus.cs (hunk 16 above)
Comment (garayx) [reply src/Raven.Server/Rachis/RachisConsensus.cs]: removed

File: src/Raven.Client/Documents/Operations/Operation.cs (hunk 13 above)
Comment (garayx) [reply src/Raven.Client/Documents/Operations/Operation.cs:200]: done

File: src/Raven.Client/Documents/DocumentStore.cs [hunk 17]
```diff
@@ -268,21 +268,31 @@ public override IDisposable DisableAggressiveCaching(string database = null)
         /// <summary>
         /// Subscribe to change notifications from the server
         /// </summary>
-        public override IDatabaseChanges Changes(string database = null)
+        public override IDatabaseChanges Changes(string database = null, string nodeTag = null)
```
Comment (ppekrol) [inline src/Raven.Client/Documents/DocumentStore.cs]: You need to create a separate `Changes` method overload to not brake the binary compatibility

File: src/Raven.Client/Documents/Operations/MaintenanceOperationExecutor.cs [hunk 18]
```diff
@@ -75,7 +75,7 @@ public async Task<Operation> SendAsync(IMaintenanceOperation<OperationIdResult>
                 var command = operation.GetCommand(RequestExecutor.Conventions, context);
 
                 await RequestExecutor.ExecuteAsync(command, context, sessionInfo: null, token: token).ConfigureAwait(false);
-                return new Operation(RequestExecutor, () => _store.Changes(_databaseName), RequestExecutor.Conventions, command.Result.OperationId);
+                return new Operation(RequestExecutor, () => _store.Changes(_databaseName), RequestExecutor.Conventions, command.Result.OperationId, command.SelectedNodeTag ?? command.Result.OperationNodeTag);
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Operations/MaintenanceOperationExecutor.cs:78]: Shouldn't this be the other way around? First take the node tag from result, then from command?

File: src/Raven.Client/Documents/DocumentStore.cs [hunk 19]
```diff
@@ -268,21 +268,31 @@ public override IDisposable DisableAggressiveCaching(string database = null)
...
-            return _databaseChanges.GetOrAdd(database ?? Database, CreateDatabaseChanges);
+            return _databaseChanges.GetOrAdd(new RequestDestination
+            {
+                DatabaseName = database ?? Database,
+                NodeTag = nodeTag
+            }, CreateDatabaseChanges);
         }
 
-        protected virtual IDatabaseChanges CreateDatabaseChanges(string database)
+        internal virtual IDatabaseChanges CreateDatabaseChanges(RequestDestination node)
```
Comment (ppekrol) [inline src/Raven.Client/Documents/DocumentStore.cs]: rename `RequestDestination` to `DatabaseChangesOptions`

File: src/Raven.Client/Util/AtomicDictionary.cs [hunk 20]
```diff
@@ -7,10 +7,10 @@
 
 namespace Raven.Client.Util
 {
-    internal class AtomicDictionary<TVal> : IEnumerable<KeyValuePair<string, TVal>>
+    internal class AtomicDictionary<TVal> : IEnumerable<KeyValuePair<RequestDestination, TVal>>
```
Comment (ppekrol) [inline src/Raven.Client/Util/AtomicDictionary.cs]: should be TKey, but I do not think we need `AtomicDictionary` at this point? Are we using any of it's features? This can be just a concurrent dictionary? cc @ayende

File: src/Raven.Client/Util/AtomicDictionary.cs [hunk 21]
```diff
@@ -163,4 +166,30 @@ public IDisposable TryWithAllLocks()
             return _globalLocker.TryEnterWriteLock(TimeSpan.FromSeconds(3));
         }
     }
-}
\ No newline at end of file
+
+    internal class AtomicDictionaryComparer : IEqualityComparer<RequestDestination>
```
Comment (ppekrol) [inline src/Raven.Client/Util/AtomicDictionary.cs]: create internal comparer for DatabaseChangesOptions

**Review by ppekrol — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Operations/MaintenanceOperationExecutor.cs (hunk 18 above)
Comment (garayx) [reply src/Raven.Client/Documents/Operations/MaintenanceOperationExecutor.cs:78]: If we select node to execute => execute on this (selected) node
if we dont select => execute on command.Result.OperationNodeTag

File: src/Raven.Client/Documents/DocumentStore.cs (hunk 17 above)
Comment (garayx) [reply src/Raven.Client/Documents/DocumentStore.cs]: done

File: src/Raven.Client/Documents/DocumentStore.cs (hunk 19 above)
Comment (garayx) [reply src/Raven.Client/Documents/DocumentStore.cs]: done

File: src/Raven.Client/Util/AtomicDictionary.cs (hunk 20 above)
Comment (garayx) [reply src/Raven.Client/Util/AtomicDictionary.cs]: done

File: src/Raven.Client/Util/AtomicDictionary.cs (hunk 21 above)
Comment (garayx) [reply src/Raven.Client/Util/AtomicDictionary.cs]: done

**Review by ppekrol — APPROVED** [review]
